client/service/src/metrics.rs

```diff
@@ -0,0 +1,250 @@
+
+use prometheus_endpoint::{register, Gauge, U64, F64, Registry, PrometheusError, Opts, GaugeVec};
+use sysinfo::{get_current_pid, ProcessExt, System, SystemExt};
+use sc_telemetry::{telemetry, SUBSTRATE_INFO};
+use sc_client::ClientInfo;
+use crate::NetworkStatus;
+use sp_transaction_pool::PoolStatus;
+use sp_runtime::traits::{NumberFor, Block, SaturatedConversion, UniqueSaturatedInto};
+use netstat2::{TcpState, ProtocolSocketInfo, iterate_sockets_info, AddressFamilyFlags, ProtocolFlags};
+
+struct PrometheusMetrics {
+	block_height_number: GaugeVec<U64>,
+	ready_transactions_number: Gauge<U64>,
+	memory_usage_bytes: Gauge<U64>,
+	netstat: GaugeVec<U64>,
+	load_avg: GaugeVec<F64>,
+	cpu_usage_percentage: Gauge<F64>,
+	network_per_sec_bytes: GaugeVec<U64>,
+	database_cache: Gauge<U64>,
+	state_cache: Gauge<U64>,
+	state_db: GaugeVec<U64>,
+}
+
+impl PrometheusMetrics {
+	fn setup(registry: &Registry, name: &str, version: &str, roles: u64)
+		-> Result<Self, PrometheusError>
+	{
+        register(Gauge::<U64>::with_opts(
+            Opts::new(
+                "build_info",
+                "A metric with a constant '1' value labeled by name, version, and commit."
+            )
+                .const_label("name", name)
+                .const_label("version", version)
+		)?, &registry)?.set(1);
+		
+        register(Gauge::<U64>::new(
+            "node_roles", "The roles the node is running as",
+		)?, &registry)?.set(roles);
+		
+		Ok(Self {
+			block_height_number: register(GaugeVec::new(
+				Opts::new("block_height_number", "Height of the chain"),
+				&["status"]
+			)?, registry)?,
+			ready_transactions_number: register(Gauge::new(
+				"ready_transactions_number", "Number of transactions in the ready queue",
+			)?, registry)?,
+			memory_usage_bytes: register(Gauge::new(
+				"memory_usage_bytes", "Node memory usage",
+			)?, registry)?,
+			cpu_usage_percentage: register(Gauge::new(
+				"cpu_usage_percentage", "Node CPU usage",
+			)?, registry)?,
+			network_per_sec_bytes: register(GaugeVec::new(
+				Opts::new("network_per_sec_bytes", "Networking bytes per second"),
+				&["direction"]
+			)?, registry)?,
+			database_cache: register(Gauge::new(
+				"database_cache_bytes", "RocksDB cache size in bytes",
+			)?, registry)?,
+			state_cache: register(Gauge::new(
+				"state_cache_bytes", "State cache size in bytes",
+			)?, registry)?,
+			state_db: register(GaugeVec::new(
+				Opts::new("state_db_cache_bytes", "State DB cache in bytes"),
+				&["subtype"]
+			)?, registry)?,
+			netstat: register(GaugeVec::new(
+				Opts::new("netstat_tcp", "Current TCP connections "),
+				&["status"]
+			)?, registry)?,
+			load_avg: register(GaugeVec::new(
```

Comment:
isn't it (various averages) something that prometheus is fine calculating and showing on it's own, given it is fed regularly with instant data?

Reply:
"load average" is what system provides, to my understanding it is the lowest value we can use. one could argue that prometheus would be able to calculate the 10min/15min running out of the one-min-averages, but I suppose the system one is more correct...